Source/WebCore/editing/cocoa/AttributedString.mm

```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+inline static ParagraphStyle extractParagraphStyle(NSParagraphStyle *style, TableToIdentifierMap& tableIDs, TableBlockToIdentifierMap& tableBlockIDs, ListToIdentifierMap& listIDs)
 {
-    return makeVector(style.textBlocks, [&](NSTextBlock *block) -> std::optional<std::optional<AttributedString::TableBlockAndTableIDPair>> {
-        if (![block isKindOfClass:PlatformNSTextTableBlock])
-            return std::nullopt;
+    Vector<AttributedStringTextTableBlockID> sentTextTableBlockIDs;
+    Vector<AttributedStringTextListID> sentTextListIDs;
+    Vector<TextTableBlock> newTextTableBlocks;
+    Vector<TextTable> newTextTables;
+    Vector<ParagraphStyleTextList> newTextLists;
+    Vector<TextTab> newTextTabs;
+
+    for (NSTextList *list in style.textLists) {
+        if (![list isKindOfClass:PlatformNSTextList])
+            continue;
+
+        auto listResult = listIDs.ensure(list, [] {
+            return AttributedString::TextListID::generate();
+        });
+        auto listID = listResult.iterator->value;
+
+        if (listResult.isNewEntry) {
+            newTextLists.append(ParagraphStyleTextList {
+                listID,
+                [list markerFormat],
+                [list startingItemNumber]
+            });
+        }
+        sentTextListIDs.append(listID);
+    }
 
-        auto tableBlock = static_cast<NSTextTableBlock *>(block);
+    for (NSTextTableBlock* item in style.textBlocks) {
+        if (![item isKindOfClass:PlatformNSTextTableBlock])
+            return { };
+
+        auto tableBlock = static_cast<NSTextTableBlock *>(item);
         if (!tableBlock.table)
-            return std::nullopt;
+            return { };
 
-        auto tableBlockID = tableBlockIDs.ensure(tableBlock, [] {
+        auto tableBlockEnsureResult = tableBlockIDs.ensure(tableBlock, [&] {
             return AttributedString::TextTableBlockID::generate();
-        }).iterator->value;
+        });
+        auto tableBlockID = tableBlockEnsureResult.iterator->value;
+
+        sentTextTableBlockIDs.append(tableBlockID);
 
-        auto tableID = tableIDs.ensure(tableBlock.table, [] {
+        auto nstable = tableBlock.table;
+        auto tableEnsureResults = tableIDs.ensure(nstable, [&] {
             return AttributedString::TextTableID::generate();
-        }).iterator->value;
+        });
+        auto tableID = tableEnsureResults.iterator->value;
+
+        TextTableLayoutAlgorithm layout;
+        switch ([nstable layoutAlgorithm]) {
+        case NSTextTableAutomaticLayoutAlgorithm:
+            layout = TextTableLayoutAlgorithm::Automatic;
+            break;
+        case NSTextTableFixedLayoutAlgorithm:
+            layout = TextTableLayoutAlgorithm::Fixed;
+            break;
+        }
 
-        return std::optional { std::pair { tableBlockID, tableID } };
-    });
+        if (tableEnsureResults.isNewEntry) {
+            std::optional<RetainPtr<PlatformColor>> backgroundColor;
```

Comment:
optional RetainPtr seems like it probably has too many states, null being distinct from std::nullopt.